### ISSUE TITLE
Add analyzer to detect async void usage in MVC filter usage

### ISSUE THREAD

**Issue by davidfowl** [dotnet/aspnetcore#28217]
People aren't aware that IAsync*Filters exist and instead use the synchronous ones with async void. It's pretty common and the errors you get are bizzare. Make this an error and offer a codefix to change the synchronous filter to an async one.

**Comment by Youssef1313** [comment]
@davidfowl I'm interested to work on this. As far as I understand, a class that implements non-async filter with async-methods should be reported by the analyzer. Is that correct? Are there any other cases? Thanks.

**Comment by JunTaoLuo** [comment]
Sounds like we are unclear where/how this analyzer would ship cc @pranavkm @davidfowl we should discuss this before we can make progress on this analyzer.

**Comment by davidfowl** [comment]
@Youssef1313 are you still interested?

**Comment by Youssef1313** [comment]
@davidfowl Yup! If possible, I need some more information since my knowledge to ASP.NET Core isn't that great. This is what I understand:

- [ ] Startup.Configure/ConfigureServices

    If a class name that starts with `Startup` (e.g, `Startup`, `StartupDevelopment`, etc.) contains an `async void` `Configure` or `ConfigureServices` method, then produce a diagnostic with a fixer to turn it to `async Task`.

- [ ] MVC Controller Action methods

    If any method in a class that derives from `Controller` contains an `async void` method, then produce a diagnostic with a fixer to turn it to `async Task`.

- [ ] Razor Page handlers

    Is the link beside this point in the issue was supposed to belong to the previous point? Otherwise, I need some distinction between this point and the previous one?

- [ ] MVC filters (the non IAsync*Filters)

    Take `IActionFilter` as an example, if a class that implements it has an async `OnActionExecuting` or `OnActionExecuted`, then produce a diagnostic with no codefix since the user should reason about that.

**Comment by BrennanConroy** [comment]
Hub methods could also be included here.

**Comment by davidfowl** [comment]
Added

**Comment by davidfowl** [comment]
cc @rafikiassumani-msft can we add this to the analyzer work for .NET 7?

**Comment by davidfowl** [comment]
@JamesNK @captainsafia Can we get this one added to the list?

**Comment by mkArtakMSFT** [comment]
Guys, bringing this back from `Backlog` for you to consider this for .NET 9, as we just saw another customer hitting this here: https://github.com/dotnet/aspnetcore/issues/51702

**Comment by halter73** [comment]
When we were discussing this in our Web UI planning meeting, we debated whether it makes sense to have a blanket analyzer for all `async void` usage in Web SDK projects.

I'm sure a lot of usage is okay, but I'd argue most of it is at least a small bug in the context of an ASP.NET Core app. If the async handler is accessing any services, it's completion should be awaited by host shutdown somehow.

If we focus too much on specific problem areas like MVC actions and page handlers, we might miss other problematic issues deep in helper logic. And if you're convinced that your usage is okay because it doesn't rely on the HttpContext and you don't care about ODE's during shutdown (or you're really not relying on any services), you're free to suppress the analyzer.

I personally think we should make our job easier implementing and maintaining the analyzer by flagging any `async void` methods in a WebSDK project.

**Comment by ericstj** [comment]
Just spent time debugging accidental async void misuse in system.commandline: https://github.com/dotnet/command-line-api/blob/77b277a9dc0981b2bbeaebb3cb17baded0c8cc51/src/System.CommandLine/Command.cs#L107

[Sample syntax](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYBGAWACgKAGAAgpoDoAVAC1gCGhAJZ4A5gG5GLdjQCsUhtJoBOABQAiQDwbgEH2NASkWNG3GAGcMnALJqKADlZqIwAFYwwGVhH2sAvAD5WRlYQ0NYAbyCGMJiQihV2JE4AERgAGwEATzUaQ2DYsI51DQAJdLSIA0UCgF88pQZTC2tbBydXd08INFYAYQE8MHKBDGEIPG4IAGsYPFYAfR8AqILI/IK4hIok1Izs3OqNuNVNMrSKqvWwuqMmGiR2KlYmjEZwq4oAZlkHihRWGwcKgAHmcbg8gQEHjGeCWgSKmiBl2ioQ+3w4v3+gLkoI6Hh6/UGw1G40mMwIiUCADN8GA4bJihQ5MiwowakA===)
